### PR TITLE
Feature/upcoming deadlines rest api

### LINKED ISSUE #219: Show time remaining for deadlines & todo-cards
There should be a time remaining field in todo cards to see deadlines easier.

Joey: can work on it after vacation(mid aug) if nobody has time

### LINKED ISSUE #329: Sort by of the TODO deck broken
# Bug report

## Expected Behavior
Sorting functions should sort as expected, e.g. by date or marking needed.

## Actual Behavior
image

## Steps to Reproduce the Problem
1. Goto home
2. Switch the sorting options of the TODO deck

### PR DESCRIPTION
**Pull request by JowehL**
## Description
Implemented deadline deck to use repeatedly.
Deadlines and todo cards now show time remaining depending on the largest factor(Days, Hours or Minutes)

## Summary of changes
- Fixes #219 
- Fixes #329

### CODE REVIEW

File: src/vue/src/components/assets/DeadlineDeck.vue [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            <b-link tag="b-button" :to="assignmentRoute(d.course.id, d.id, d.journal)">
+                <todo-card :deadline="d"/>
+            </b-link>
+        </div>
+
+        <div v-if="$route.params.cID" v-for="(d, i) in computedDeadlines" :key="i">
+            <b-link tag="b-button" :to="assignmentRoute(d.course.id, d.id, d.journal)">
+                <todo-card :deadline="d"/>
+            </b-link>
+        </div>
```
Comment (Larspolo) [inline src/vue/src/components/assets/DeadlineDeck.vue]: Why to of these, cant you just do `$route.path == '/Home' && $route.params.cID`

File: src/vue/src/views/AssignmentsOverview.vue [hunk 2]
```diff
@@ -106,7 +106,7 @@ export default {
             }
 
             function compareMarkingNeeded (a, b) {
-                return self.compare(a.stats.needs_marking, b.stats.needs_marking)
+                return self.compare(a.stats.needs_marking + a.stats.unpublished, b.stats.needs_marking + b.stats.unpublished)
```
Comment (Larspolo) [inline src/vue/src/views/AssignmentsOverview.vue]: does this still work for students?

File: src/vue/src/views/Course.vue [hunk 3]
```diff
@@ -25,22 +25,8 @@
         </b-button>
 
         <h3 slot="right-content-column">To Do</h3>
+        <deadline-deck slot="right-content-column" :deadlines="deadlines"></deadline-deck>
```
Comment (Larspolo) [inline src/vue/src/views/Course.vue]: `<deadline-deck slot="right-content-column" :deadlines="deadlines"/>`*

File: src/vue/src/views/Home.vue [hunk 4]
```diff
@@ -24,21 +24,7 @@
...
-                <option value="sortNeedsMarking">Sort by marking needed</option>
-            </b-form-select>
-        </b-card>
-
-        <div v-for="(d, i) in computedDeadlines" :key="i" slot="right-content-column">
-            <b-link tag="b-button" :to="assignmentRoute(d.course.id, d.id, d.journal)">
-                <todo-card :deadline="d"/>
-            </b-link>
-        </div>
+        <deadline-deck slot="right-content-column" :deadlines="deadlines"></deadline-deck>
```
Comment (Larspolo) [inline src/vue/src/views/Home.vue]: `<deadline-deck slot="right-content-column" :deadlines="deadlines"/>`*

File: src/vue/src/components/assets/DeadlineDeck.vue [hunk 5]
```diff
@@ -0,0 +1,94 @@
+<template>
+    <div>
+        <!-- TODO: This seems like an inappropriate permission check. Will have to be reconsidered in the rework. -->
+        <b-card v-if="$hasPermission('can_add_course')"
```
Comment (mjvkeulen) [inline src/vue/src/components/assets/DeadlineDeck.vue]: Line 4 and 5 could be combined as they are < 80 chars?

File: src/vue/src/components/assets/DeadlineDeck.vue [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                <todo-card :deadline="d"/>
+            </b-link>
+        </div>
+
+    </div>
+</template>
+
+<script>
+import todoCard from '@/components/assets/TodoCard.vue'
+import todoSquare from '@/components/assets/TodoSquare.vue'
```
Comment (mjvkeulen) [inline src/vue/src/components/assets/DeadlineDeck.vue]: Expecting newline?

File: src/vue/src/components/assets/DeadlineDeck.vue [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                params: {
+                    cID: cID,
+                    aID: aID
+                }
+            }
+
+            if (this.$hasPermission('can_view_assignment_participants', 'assignment', String(aID))) {
+                route.name = 'Assignment'
+                return route
+            } else {
```
[the remark this answers is not in the document]
Comment (JowehL) [reply src/vue/src/components/assets/DeadlineDeck.vue]: True, I will remove the else tonight

**Pull request merged**